### PR TITLE
Subset of LSP - Part 1

### PR DESCRIPTION
**Pull request by DmytroMitin**
### Pull Request Description

For each of the below-listed messages
```
General: initialize
General: initialized
General: shutdown
General: exit
```
they are implemented in the gateway. Corresponding akka messages between the gateway and the language server are implemented (with a stub response). 

#392 

### Important Notes
N/A

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/gateway/src/main/resources/application.conf [hunk 1]
```diff
@@ -4,6 +4,8 @@ gateway {
     port = 30000
     route = ""
     timeout = 5
+    bindingTimeout = 5
```
Comment (iamrecursion) [inline engine/gateway/src/main/resources/application.conf]: Please change these names to be more descriptive if you can. It gives no indication what the units might be. Something like `timeoutSecs` would be much better (or whatever the unit actually is). If these names are limited for some reason, please add comments to the file indicating their units.

File: engine/gateway/src/main/scala/org/enso/Gateway.scala [hunk 2]
```diff
@@ -16,32 +16,60 @@ import org.enso.gateway.protocol.response.result.{
   */
 class Gateway(languageServer: ActorRef) extends Actor with ActorLogging {
   override def receive: Receive = {
-    case Requests.Initialize(Id.Number(id), _) =>
+    case Requests.Initialize(id, params) =>
       val msg = "Gateway: Initialize received"
       log.info(msg)
-      languageServer ! LanguageServer.Initialize(id, sender())
+      languageServer ! languageserver.Requests.Initialize(
```
Comment (iamrecursion) [inline engine/gateway/src/main/scala/org/enso/Gateway.scala]: The lower-case `languageserver` here is a strange naming choice. Why?

File: engine/gateway/src/main/scala/org/enso/gateway/protocol/Requests.scala [hunk 3]
```diff
@@ -1,14 +1,13 @@
 package org.enso.gateway.protocol
 
 import org.enso.gateway.protocol.request.Params
+import org.enso.gateway.protocol.request.Params.{InitializeParams, VoidParams}
 
-/** Parent trait for requests extractor objects. */
-sealed trait Requests {
+/** Parent trait for request extractor objects. */
```
Comment (iamrecursion) [inline engine/gateway/src/main/scala/org/enso/gateway/protocol/Requests.scala]: This should explain what request extractor objects are and what they do.

File: engine/gateway/src/main/scala/org/enso/Gateway.scala (hunk 2 above)
Comment (DmytroMitin) [reply engine/gateway/src/main/scala/org/enso/Gateway.scala]: `languageserver` is a package, `org.enso.languageserver`.

Comment (iamrecursion) [reply engine/gateway/src/main/scala/org/enso/Gateway.scala]: I know. Is it _necessary_ to use a package-qualified import here?

Comment (DmytroMitin) [reply engine/gateway/src/main/scala/org/enso/Gateway.scala]: I have to distinguish `org.enso.gateway.protocol.Requests` and `org.enso.languageserver.Requests`. I will think about naming. Maybe I'll use "rename on import" syntax. Thanks.

Comment (DmytroMitin) [reply engine/gateway/src/main/scala/org/enso/Gateway.scala]: Fixed. Thanks.

**Comment by DmytroMitin** [comment]
@iamrecursion 
>     * I would like to see some end-to-end tests of the init/shutdown workflow.
I added tests
https://github.com/luna/enso/blob/wip/dmytro/starting-subset-of-lsp-part1/engine/gateway/src/test/scala/org/enso/gateway/GatewayMessageSpec.scala
https://github.com/luna/enso/blob/wip/dmytro/starting-subset-of-lsp-part1/engine/language-server/src/test/scala/org/enso/languageserver/LanguageServerSpec.scala
Is this what you meant?

File: engine/gateway/src/main/scala/org/enso/gateway/protocol/CodeActionKind.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+import io.circe.{Decoder, Encoder}
+
+/** Kind of
+  * [[org.enso.gateway.protocol.request.clientcapabilities.textdocument.CodeAction]].
+  *
+  * Used also in
+  * [[org.enso.gateway.protocol.response.result.servercapabilities.CodeActionProvider]].
+  */
+sealed abstract class CodeActionKind(val value: String)
```
Comment (kustosz) [inline engine/gateway/src/main/scala/org/enso/gateway/protocol/CodeActionKind.scala]: Are these implemented? Are they part of the task specification? What is the purpose of adding this code?

File: engine/gateway/src/main/scala/org/enso/gateway/protocol/request/clientcapabilities/common/SymbolKind.scala [hunk 5]
```diff
@@ -0,0 +1,118 @@
+package org.enso.gateway.protocol.request.clientcapabilities.common
+
+import io.circe.Decoder
+
+/** A symbol kind. */
+sealed abstract class SymbolKind(value: Int)
```
Comment (kustosz) [inline engine/gateway/src/main/scala/org/enso/gateway/protocol/request/clientcapabilities/common/SymbolKind.scala]: Again, why add it all at this stage? We don't even support half of these (what's an enum?).

File: engine/gateway/src/main/scala/org/enso/gateway/protocol/CodeActionKind.scala (hunk 4 above)
Comment (DmytroMitin) [reply engine/gateway/src/main/scala/org/enso/gateway/protocol/CodeActionKind.scala]: `ClientCapabilities` and `ServerCapabilities` are JSON-RPC parameters of request `initialize`.

File: engine/gateway/src/main/scala/org/enso/gateway/protocol/request/clientcapabilities/common/SymbolKind.scala (hunk 5 above)
Comment (DmytroMitin) [reply engine/gateway/src/main/scala/org/enso/gateway/protocol/request/clientcapabilities/common/SymbolKind.scala]: `ClientCapabilities` and `ServerCapabilities` are JSON-RPC parameters of request/response `initialize`.

**Pull request merged**